python/pyspark/streaming/dstream.py

```diff
@@ -51,125 +78,170 @@ class DStream:
      - A function that is used to generate an RDD after each time interval
     """
 
-    def __init__(self, jdstream, ssc, jrdd_deserializer):
+    def __init__(
+        self,
+        jdstream: JavaObject,
+        ssc: "StreamingContext",
+        jrdd_deserializer: "Serializer",
+    ) -> None:
         self._jdstream = jdstream
         self._ssc = ssc
-        self._sc = ssc._sc
+        self._sc = ssc._sc  # type: ignore[attr-defined]
         self._jrdd_deserializer = jrdd_deserializer
         self.is_cached = False
         self.is_checkpointed = False
 
-    def context(self):
+    def context(self) -> "StreamingContext":
         """
         Return the StreamingContext associated with this DStream
         """
         return self._ssc
 
-    def count(self):
+    def count(self) -> "DStream[int]":
         """
         Return a new DStream in which each RDD has a single element
         generated by counting each RDD of this DStream.
         """
         return self.mapPartitions(lambda i: [sum(1 for _ in i)]).reduce(operator.add)
 
-    def filter(self, f):
+    def filter(self, f: Callable[[T_co], bool]) -> "DStream[T_co]":
```

Comment:
In general, we should avoid using covariant types in arguments, unless necessary. See how it is done in `pyspark.rdd.RDD`.